docs/abi.md

```diff
@@ -190,6 +242,22 @@ If creating the specified Channel would violate
 - `param[3]: usize`: Label size in bytes
 - `result[0]: u32`: Status of operation
 
+### `channel_create_with_privilege`
+
+Variant of `channel_create` that allows a Node to perform a channel creation
+operation that requires use of its downgrade privilege to satisfy
+[information flow control](/docs/concepts.md#information-flow-control) flows-to
+checks.
+
+The privilege to use is provided as a serialized
+[`Label`](/oak/proto/label.proto) protobuf message; if the calling Node does not
+have this privilege, `ERR_INVALID_ARGS` is returned.
+
+- `param[0]` to `param[3]`: As for `channel_create`
+- `param[4]: usize`: Privilege to use in downgrade, as a serialized `Label`
+- `param[5]: usize`: Privilege label size in bytes
+- `result[0]: u32`: Status of operation
+
 ### `channel_close`
```

Comment:
Is there never a case where channel close needs to use privilege? Closing a channel must be visible to someone. Is it that only nodes with the write half have the ability to close a channel?

If nodes with the read half of a channel can also close it, this could have a an information flow violation.

We should be doing LabelOf(node) flowsTo LabelOf(channel) (if we aren't already) during channel_close. There can be a downgrading option.

Reply:
At the moment `channel_close` doesn't do any sort of label check, so there's no `..._with_privilege` variant. 

Also, at the moment a `channel_close` is indeed potentially visible to someone &ndash; someone writing to the channel gets an "orphaned channel" return value to indicate that the write is pointless. That's definitely a possible information flow violation: discussion is in #763, where I'm concerned about the pragmatic effects of never being able to garbage collect state (nor cope with failure in a future distributed Runtime).

BTW, if #763 were to be implemented so there's no visible effects from a `channel_close`, would it still be the case that `channel_close` would need a flows-to check?  (I'm assuming not, but there may well be something I've missed.)

Reply:
With the current implementation, there does need to be a label check. The other safe option is to cut the channel_close call. 

It does also make sense to me that you would want to garbage collect unused channels, and that in a distributed application, some nodes might want to shut down before the whole application does. (There will be timing channels through channel_close, but I think it makes sense to treat these as out-of-scope in the near/medium-term). 

In #763, I think I missed the proposal to implement this without visible effects. I think I don't get what you mean. I think if there is no visible effect from the channel_close, then channel_close doesn't do much.

Reply:
Sorry, I probably wasn't clear enough.  

At the moment, if node B does `channel_close` on the (only) read handle for a channel, then a writer (say node A) gets a notification that the channel is orphaned.  This allows node A to take notice of the write failure and clean up any of its associated state.

#763 suggests removing the orphaned channel notification, in order to remove the info leak back from B to A.  With #763, node A can't tell that it's writing messages to a void, and can't clean up any of its own state that's associated with the channel and node B.

However, even with #763 there's still a use in doing `channel_close`: it allows the Runtime to clear up state associated with the channel.  When node A writes a message to the channel, the Runtime can observe that there is no possible way of any node reading that message, and so it can just drop the newly written message.  This stops the internal representation of the channel ballooning with queued messages that can never be retrieved.  In other words, the closure of the channel isn't visible to other nodes, but *is* visible to the Runtime.

Does that make more sense?

Reply:
This does make more sense, thanks. I think you are right that there is no leak to A with what you proposed, and it also sounds useful.

Reply:
@daviddrysdale note that if we do keep `channel_close`, even if it only affects the Runtime (by effectively skipping label checks), it means that the introspection visualization now again may depend on private data, so we would not be able to keep it in "production". Perhaps we should only allow public nodes to call `channel_close`, similar to what we are already proposing in #1143 ?